### ISSUE TITLE
GAT + enum: incorrect unused lifetime error

### ISSUE THREAD

**Issue by dhardy** [rust-lang/rust#69184]
This is a follow up on #67089, which now works. The following doesn't ([play link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d6570106bebc880bc301f30de216b611)):
```rust
#![feature(generic_associated_types)]

trait A {
    type B<'a>;
    
    fn make_b<'a>(&'a self) -> Self::B<'a>;
}

struct S {}

impl A for S {
    type B<'a> = &'a S;
    fn make_b<'a>(&'a self) -> &'a Self {
        self
    }
}

enum E<'a> {
    S(<S as A>::B<'a>),
}
```
Error is:
```
18 | enum E<'a> {
   |        ^^ unused parameter
```
however, this lifetime clearly is used and is required.

**Comment by Centril** [comment]
The variance code in `src/librustc_typeck/variance/constraints.rs` is:
```rust
            ty::Projection(ref data) => {
                let tcx = self.tcx();
                self.add_constraints_from_trait_ref(current, data.trait_ref(tcx), variance);
            }
```
It seems to me that this is disregarding the substitutions for `::B<'a>` if I'm understanding what `.trait_ref(tcx)` does here wrt. the substitutions (this pretty poorly documented right now).

cc @matthewjasper

**Comment by dhardy** [comment]
BTW how long do we need to keep this warning around? I understand that this is basically a given when using nightly features. (It puts me off using it in my normal build path — but presumably that's exactly what is wanted, in order to detect bugs like the one reported here.)
```
warning: the feature `generic_associated_types` is incomplete and may cause the compiler to crash
```

**Comment by Centril** [comment]
We only use `INCOMPLETE_FEATURES` for those that have been ICE-prone in the past or are generally not in a "considered implemented modulo bugs" state. It's quite a small set of features that are incomplete like this. GATs specifically are to my knowledge nowhere near ready to be "implemented modulo bugs".

**Comment by dhardy** [comment]
On this topic, the following builds,
```rust
trait Foo {}
impl<'a, T: A> Foo for T
where <T as A>::B<'a>: std::fmt::Debug
{}
```
but the bound `<T as A>::B<'_>: std::fmt::Debug` fails with:
```
error[E0637]: `'_` cannot be used here

I *think* this should work?
```

**Comment by dhardy** [comment]
Also, the following fails:
```rust
trait Foo {}
impl<'a, T: A> Foo for T
where <T as A>::B<'a>: 'static
{}
```
with:
```
error[E0310]: the associated type `<T as A>::B<'_>` may not live long enough
```
This bound may not make much sense (since there is another lifetime parameter), but I believe it should be valid to specify here.

**Comment by Dirbaio** [comment]
I'm running into a similar issue which I believe is the same as this: [(play)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7ded87a74db2a8b0e42b292bd66ef1a8)

```rust

trait Bar{ 
    type Foo<'a>: 'a;
}

struct What<'a, T: Bar> {
    foo: T::Foo<'a>,
}
```
